src/main/kotlin/com/clouway/bankapp/adapter/gae/JsonEntityHelpers.kt

```diff
@@ -0,0 +1,56 @@
+package com.clouway.bankapp.adapter.gae
+
+import com.clouway.bankapp.core.*
+import com.google.appengine.api.datastore.DatastoreService
+import com.google.appengine.api.datastore.Entity
+import com.google.appengine.api.datastore.EntityNotFoundException
+import com.google.appengine.api.datastore.Key
+import com.google.appengine.api.memcache.MemcacheService
+import java.time.LocalDateTime
+import java.time.ZoneId
+import java.util.*
+
+/**
+ * @author Tsvetozar Bonev (tsbonev@gmail.com)
+ */
+fun <T> MemcacheService.get(key: String, typeOfT: Class<T>, transformer: JsonTransformerWrapper): Optional<T> {
+    val entity = this.get(key) ?: return Optional.empty()
+    return Optional.of(transformer.fromJson(entity as String, typeOfT))
+}
+
+fun <T> DatastoreService.get(key: Key, typeOfT: Class<T>, transformer: JsonTransformerWrapper): Optional<T> {
+    return try {
+        val entity = this.get(key)
+        val entityContent = entity.properties["content"] as String
+        Optional.of(transformer.fromJson(entityContent, typeOfT))
+    } catch (e: EntityNotFoundException) {
+        Optional.empty()
+    }
+}
+
+fun MemcacheService.putJson(key: String, obj: Any, transformer: JsonTransformerWrapper) {
+    val objToJson = transformer.toJson(obj)
+    this.put(key, objToJson)
+}
+
+fun DatastoreService.putJson(key: Key, obj: Any, transformer: JsonTransformerWrapper) {
+    val objToJson = transformer.toJson(obj)
+    val entity = Entity(key)
+    attachObjectInformation(entity, obj)
+
+    entity.setProperty("content", objToJson)
+    this.put(entity)
+}
+
+fun LocalDateTime.toUtilDate(): Date{
+    return Date.from(this.atZone(ZoneId.systemDefault()).toInstant())
+}
+
+private fun attachObjectInformation(entity: Entity, obj: Any){
+
+    when(obj){
+        is User -> entity.setProperty("username", obj.username)
+        is Session -> entity.setProperty("expiresOn", obj.expiresOn.toUtilDate())
+        is Transaction -> entity.setProperty("userId", obj.userId)
+    }
+}
```

Comment:
here is an exmple of how we do the things
```kotlin
class TypedEntity(private val entity: Entity) {
  private val gson = Gson()
  private val indexWriter = IndexWriterImpl()

  val key = entity.key

  fun double(key: String): Double {
    val value = entity.getProperty(key)
    return value as? Double ?: 0.0
  }

  fun string(key: String): String {
    val value = entity.getProperty(key)
    return value as? String ?: (value as Text).value
  }

  fun stringOr(key: String, defaultValue: String): String {
    val value = entity.getProperty(key) ?: return defaultValue
    return value as? String ?: (value as Text).value
  }

  fun stringOrNull(key: String): String? {
    val value = entity.getProperty(key) ?: return null
    return value as? String ?: (value as Text).value
  }

  fun longValue(key: String): Long {
    return entity.getProperty(key) as Long
  }

  fun longValueOrNull(key: String): Long? {
    val value = entity.getProperty(key)
    return if (value != null) value as Long else null
  }

  fun longValueOr(key: String, defaultValue: Long): Long {
    val value = entity.getProperty(key) ?: return defaultValue
    return value as Long
  }

  fun booleanValue(key: String): Boolean {
    val value = entity.getProperty(key) ?: false
    return value as Boolean
  }

  fun booleanValueOr(key: String, default: Boolean): Boolean {
    val value = entity.getProperty(key) ?: return default
    return value as Boolean
  }

  fun date(key: String): Date? {
    return if (entity.getProperty(key) == null) null else entity.getProperty(key) as Date
  }

  fun dateValue(key: String): Date = entity.getProperty(key) as Date

  fun dateValueOr(key: String, defaultValue: Date): Date {
    return if (entity.getProperty(key) == null) defaultValue else entity.getProperty(key) as Date
  }

  fun dateTimeValueOrNull(key: String): LocalDateTime? {
    val date = if (entity.getProperty(key) == null) return null else entity.getProperty(key) as Date
    return LocalDateTime.ofInstant(date.toInstant(), ZoneOffset.UTC)
  }

  fun setUnindexedDateTimeValue(key: String, dateTime: LocalDateTime) {
    val date = Date(dateTime.toInstant(ZoneOffset.UTC).toEpochMilli())
    entity.setUnindexedProperty(key, date)
  }

  fun setIndexedDateTimeValue(key: String, dateTime: LocalDateTime) {
    val date = Date(dateTime.toInstant(ZoneOffset.UTC).toEpochMilli())
    entity.setIndexedProperty(key, date)
  }

  fun byteArrayOr(propertyName: String, orValue: ByteArray): ByteArray {
    val result = entity.getProperty(propertyName) ?: return orValue
    return (result as Blob).bytes ?: orValue
  }

  fun <T> jsonValueOr(property: String, type: Type, defaultValue: T): T {
    if (!entity.hasProperty(property)) return defaultValue

    val json = entity.getProperty(property) as Text
    return gson.fromJson<T>(json.value, type) ?: defaultValue
  }

  fun <T> setJsonValue(property: String, value: T) {
    val textValue = gson.toJson(value)
    entity.setUnindexedProperty(property, Text(textValue))
  }

  @Suppress("UNCHECKED_CAST")
  fun <T> list(propertyName: String): List<T> {
    if (!entity.hasProperty(propertyName)) return listOf()

    val value = (entity.getProperty(propertyName) ?: return listOf()) as? kotlin.collections.Collection<*>
            ?: throw IllegalStateException(String.format("Property '%s' of entity '%s' is not of type collection.", propertyName, entity.kind))

    if (value is List<*>) return value as List<T>
    return (value as kotlin.collections.Collection<T>).toList()
  }

  @Suppress("UNCHECKED_CAST")
  fun <T> set(propertyName: String): Set<T> {
    if (!entity.hasProperty(propertyName)) return emptySet()

    val o = (entity.getProperty(propertyName) ?: return setOf()) as? kotlin.collections.Collection<*>
            ?: throw IllegalStateException(String.format("Property '%s' of raw '%s' is not of type collection.", propertyName, entity.kind))

    if (o is kotlin.collections.Set<*>) return o as Set<T>
    return (o as kotlin.collections.Collection<T>).toSet()
  }

  fun setUnindexedProperty(key: String, value: Any) {

    if (value is String && value.toString().toByteArray(Charsets.UTF_8).size >= 1500) {
      entity.setUnindexedProperty(key, Text(value.toString()))
      return
    }
    entity.setUnindexedProperty(key, value)
  }

  fun setSearchIndexProperty(values: List<String>, wholeValues: List<String> = emptyList(), propertyName: String = "searchIndex") {
    entity.setIndexedProperty(propertyName, indexWriter.createIndex(*values.toTypedArray()).plus(wholeValues))
  }

  //Datastore has a limitation of 1500 bytes for a field.
  //If the provided field is a string with more than 1500 bytes it will be unindexed.
  fun setIndexedProperty(key: String, value: Any?) {
    if (value is String && value.toString().toByteArray(Charsets.UTF_8).size >= 1500) {
      entity.setUnindexedProperty(key, Text(value.toString()))
      return
    }
    entity.setIndexedProperty(key, value)
  }

  fun raw(): Entity = entity

}

```

Reply:
Implemented TypedEntity in Datastore Repositories